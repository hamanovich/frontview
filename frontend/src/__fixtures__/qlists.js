export const qlist = {
  _id: '5bd6e86025ccd523e4595ad4',
  questions: ['5bd2dedb23c8a03cc87f2fbc'],
  title: 'Js core',
  notes: 'only questions related to core js',
  author: '5bd19389c19fbb1840e95bcd',
  slug: 'js-core',
};

export const qlists = [qlist];
