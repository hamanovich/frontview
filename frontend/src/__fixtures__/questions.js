export const question = {
  _id: '5bd2dedb23c8a03cc87f2fbc',
  votes: {
    like: ['5bd19389c19fbb1840e95bcd'],
    dislike: [],
  },
  skill: ['CSS'],
  level: ['Middle'],
  answers: [],
  question: 'Test question',
  practice: 'practice',
  answer: 'test',
  author: '5bd2ce32cc8cfa2208379ad5',
  isVerified: false,
  lastModified: '2018-10-26T09:31:07.141Z',
  created: '2018-10-26T09:31:07.383Z',
  slug: 'test',
};

export const questions = [question];
