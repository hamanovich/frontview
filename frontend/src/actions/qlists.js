import api from '../api';

import { QLISTS_ADD, QLIST_ADD, QLIST_GET, QLIST_ADD_QUESTION, QLIST_REMOVE } from './types';

import { addQuestions } from './questions';

export const addQlists = qlists => ({
  type: QLISTS_ADD,
  qlists,
});

export const qlistAdded = qlist => ({
  type: QLIST_ADD,
  qlist,
});

export const qlistGot = qlist => ({
  type: QLIST_GET,
  qlist,
});

export const qlistQuestionAdded = qlist => ({
  type: QLIST_ADD_QUESTION,
  qlist,
});

export const qlistRemoved = qlist => ({
  type: QLIST_REMOVE,
  qlist,
});

export const qlistAdd = qlist => dispatch =>
  api.qlists.add(qlist).then(qlist => dispatch(qlistAdded(qlist)));

export const qlistAddQuestion = (qlist, question) => dispatch => () =>
  api.qlists.addQuestion(qlist, question).then(qlist => dispatch(qlistQuestionAdded(qlist)));

export const getQLists = username => dispatch =>
  api.qlists.getByAuthor(username).then(qlists => dispatch(addQlists(qlists)));

export const getQListQuestions = (username, slug) => dispatch =>
  api.qlists.getQListQuestions(username, slug).then(qlist => {
    dispatch(qlistGot(qlist));
    dispatch(addQuestions(qlist.questions));
  });

export const removeQList = _id => dispatch =>
  api.qlists.remove(_id).then(qlist => dispatch(qlistRemoved(qlist)));
