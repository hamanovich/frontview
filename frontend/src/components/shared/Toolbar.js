import React from 'react';
import { arrayOf, shape, func, string } from 'prop-types';
import { connect } from 'react-redux';
import { LinkContainer } from 'react-router-bootstrap';
import FontAwesome from 'react-fontawesome';
import map from 'lodash/map';

import Button from 'react-bootstrap/Button';
import ButtonGroup from 'react-bootstrap/ButtonGroup';
import ButtonToolbar from 'react-bootstrap/ButtonToolbar';
import DropdownButton from 'react-bootstrap/DropdownButton';
import Dropdown from 'react-bootstrap/Dropdown';

import { voteQuestion } from '../../actions/questions';
import { qlistAddQuestion } from '../../actions/qlists';

import { QuestionType, QListType } from '../../propTypes';

export const Toolbar = ({ user, question, voteQuestion, qlistAddQuestion, qlists }) => (
  <ButtonToolbar>
    <ButtonGroup size="sm">
      <Button
        variant="success"
        active={question.votes.like.includes(user._id)}
        onClick={voteQuestion(question, 'like', user._id)}>
        <FontAwesome name="thumbs-up" /> {question.votes.like.length}
      </Button>

      <Button
        variant="danger"
        active={question.votes.dislike.includes(user._id)}
        onClick={voteQuestion(question, 'dislike', user._id)}>
        <FontAwesome name="thumbs-down" /> {question.votes.dislike.length}
      </Button>
    </ButtonGroup>

    <ButtonGroup size="sm">
      <DropdownButton size="sm" variant="info" title={<FontAwesome name="star" />} id="qlist">
        {map(qlists, (qlist, index) => (
          <Dropdown.Item
            eventKey={index}
            key={qlist._id}
            onClick={qlistAddQuestion(qlist, question)}>
            {qlist.title}
            {map(qlist.questions, q => q._id).includes(question._id) && (
              <FontAwesome name="check" />
            )}
          </Dropdown.Item>
        ))}
        {qlists.length > 0 && <Dropdown.Divider />}
        <LinkContainer to="/me/qlist/create">
          <Dropdown.Item>Create a new QList</Dropdown.Item>
        </LinkContainer>
      </DropdownButton>
    </ButtonGroup>
  </ButtonToolbar>
);

Toolbar.propTypes = {
  user: shape({
    _id: string,
  }).isRequired,
  question: QuestionType.isRequired,
  qlists: arrayOf(QListType).isRequired,
  qlistAddQuestion: func.isRequired,
  voteQuestion: func.isRequired,
};

export default connect(
  null,
  {
    voteQuestion,
    qlistAddQuestion,
  },
)(Toolbar);
