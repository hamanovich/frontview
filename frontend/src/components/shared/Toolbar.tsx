import React, { FunctionComponent } from 'react';
import { connect } from 'react-redux';
import { LinkContainer } from 'react-router-bootstrap';
import FontAwesome from 'react-fontawesome';

import Button from 'react-bootstrap/Button';
import ButtonGroup from 'react-bootstrap/ButtonGroup';
import ButtonToolbar from 'react-bootstrap/ButtonToolbar';
import DropdownButton from 'react-bootstrap/DropdownButton';
import Dropdown from 'react-bootstrap/Dropdown';

import { voteQuestion } from '../../actions/questions';
import { qlistAddQuestion } from '../../actions/qlists';

import { Question } from '../../propTypes/QuestionType';
import { QList } from '../../propTypes/QListType';
import { ToolbarProps } from './models';

export const Toolbar: FunctionComponent<ToolbarProps> = ({
  user,
  question,
  voteQuestion,
  qlistAddQuestion,
  qlists,
}) => (
  <ButtonToolbar>
    <ButtonGroup size="sm">
      <Button
        variant="success"
        active={question.votes.like.includes(user._id)}
        onClick={voteQuestion(question, 'like', user._id)}>
        <FontAwesome name="thumbs-up" /> {question.votes.like.length}
      </Button>

      <Button
        variant="danger"
        active={question.votes.dislike.includes(user._id)}
        onClick={voteQuestion(question, 'dislike', user._id)}>
        <FontAwesome name="thumbs-down" /> {question.votes.dislike.length}
      </Button>
    </ButtonGroup>

    <ButtonGroup size="sm">
      <DropdownButton
        size="sm"
        variant="info"
        title={<FontAwesome name="star" />}
        id="qlist">
        {qlists.map((qlist: QList, index: any) => (
          <Dropdown.Item
            eventKey={index}
            key={qlist._id}
            onClick={qlistAddQuestion(qlist, question)}>
            {qlist.title}
            {qlist.questions
              .map((q: Question) => q._id)
              .includes(question._id) && <FontAwesome name="check" />}
          </Dropdown.Item>
        ))}
        {qlists.length > 0 && <Dropdown.Divider />}
        <LinkContainer to="/me/qlist/create">
          <Dropdown.Item>Create a new QList</Dropdown.Item>
        </LinkContainer>
      </DropdownButton>
    </ButtonGroup>
  </ButtonToolbar>
);

export default connect(
  null,
  {
    voteQuestion,
    qlistAddQuestion,
  },
)(Toolbar);
