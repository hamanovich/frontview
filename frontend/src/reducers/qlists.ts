import map from 'lodash/map';

import {
  QLISTS_ADD,
  QLIST_ADD,
  QLIST_GET,
  QLIST_ADD_QUESTION,
  QLIST_REMOVE,
} from '../actions/types';
import { QList } from '../propTypes/QListType';
import { QListActionTypes } from '../actions/qlists';

const initialState: QList[] = [];

export default (state = initialState, action: QListActionTypes): QList[] => {
  switch (action.type) {
    case QLISTS_ADD:
      return action.qlists;

    case QLIST_ADD:
      return [...state, action.qlist];

    case QLIST_GET: {
      const gotIndex = state.findIndex(qlist => qlist._id === action.qlist._id);

      if (gotIndex > -1) {
        return map(state, qlist => {
          if (qlist._id === action.qlist._id) {
            return action.qlist;
          }

          return qlist;
        });
      }

      return [action.qlist];
    }

    case QLIST_ADD_QUESTION: {
      const qlistIndex = state.findIndex(
        qlist => qlist._id === action.qlist._id,
      );

      if (qlistIndex > -1) {
        return map(state, qlist => {
          if (qlist._id === action.qlist._id) {
            return action.qlist;
          }

          return qlist;
        });
      }

      return [action.qlist];
    }

    case QLIST_REMOVE: {
      const removeIndex = state.findIndex(
        qlist => qlist._id === action.qlist._id,
      );

      if (removeIndex > -1) {
        return [
          ...state.slice(0, removeIndex),
          ...state.slice(removeIndex + 1),
        ];
      }

      return state;
    }

    default:
      return state;
  }
};
