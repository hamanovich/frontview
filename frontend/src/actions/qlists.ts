import { Action } from 'redux';
import { ThunkAction } from 'redux-thunk';

import api from '../api';
import { addQuestions } from './questions';
import { AppState } from '../reducers';
import { QList } from '../propTypes/QListType';
import { Question } from '../propTypes/QuestionType';
import {
  QLISTS_ADD,
  QLIST_ADD,
  QLIST_GET,
  QLIST_ADD_QUESTION,
  QLIST_REMOVE,
} from './types';

interface addQlistsAction {
  type: typeof QLISTS_ADD;
  qlists: QList[];
}

interface qlistAddedAction {
  type: typeof QLIST_ADD;
  qlist: QList;
}

interface qlistGotAction {
  type: typeof QLIST_GET;
  qlist: QList;
}

interface qlistQuestionAddedAction {
  type: typeof QLIST_ADD_QUESTION;
  qlist: QList;
}

interface qlistRemovedAction {
  type: typeof QLIST_REMOVE;
  qlist: QList;
}

export const addQlists = (qlists: QList[]) => ({
  type: QLISTS_ADD,
  qlists,
});

export const qlistAdded = (qlist: QList) => ({
  type: QLIST_ADD,
  qlist,
});

export const qlistGot = (qlist: QList) => ({
  type: QLIST_GET,
  qlist,
});

export const qlistQuestionAdded = (qlist: QList) => ({
  type: QLIST_ADD_QUESTION,
  qlist,
});

export const qlistRemoved = (qlist: QList) => ({
  type: QLIST_REMOVE,
  qlist,
});

export const qlistAdd = (
  qlist: QList,
): ThunkAction<void, AppState, null, Action<string>> => dispatch =>
  api.qlists.add(qlist).then(qlist => dispatch(qlistAdded(qlist)));

export const qlistAddQuestion = (
  qlist: QList,
  question: Question,
): ThunkAction<void, AppState, null, Action<string>> => dispatch => () =>
  api.qlists
    .addQuestion(qlist, question)
    .then(qlist => dispatch(qlistQuestionAdded(qlist)));

export const getQLists = (
  username: string,
): ThunkAction<void, AppState, null, Action<string>> => dispatch =>
  api.qlists.getByAuthor(username).then(qlists => dispatch(addQlists(qlists)));

export const getQListQuestions = (
  username: string,
  slug: string,
): ThunkAction<void, AppState, null, Action<string>> => dispatch =>
  api.qlists.getQListQuestions(username, slug).then(qlist => {
    dispatch(qlistGot(qlist));
    dispatch(addQuestions(qlist.questions));
  });

export const removeQList = (
  _id: string,
): ThunkAction<void, AppState, null, Action<string>> => dispatch =>
  api.qlists.remove(_id).then(qlist => dispatch(qlistRemoved(qlist)));

export type QListActionTypes =
  | addQlistsAction
  | qlistAddedAction
  | qlistGotAction
  | qlistQuestionAddedAction
  | qlistRemovedAction;
