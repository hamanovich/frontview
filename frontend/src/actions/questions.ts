import { Action } from 'redux';
import { ThunkAction } from 'redux-thunk';

import api from '../api';
import { AppState } from '../reducers';
import { Question } from '../propTypes/QuestionType';
import {
  QUESTIONS_ADD,
  QUESTIONS_FROM_FILE_ADD,
  QUESTION_ADD,
  QUESTION_APPROVE,
  QUESTION_EDIT,
  QUESTION_GET,
  QUESTION_REMOVE,
  VOTE_LIKE,
  VOTE_DISLIKE,
} from './types';

interface AddQuestionsAction {
  type: typeof QUESTIONS_ADD;
  questions: Question[];
  count: number;
  pages: number;
}

interface QuestionGotAction {
  type: typeof QUESTION_GET;
  question: Question;
}

interface QuestionAddedAction {
  type: typeof QUESTION_ADD;
  question: Question;
}

interface QuestionsFromFileAddedAction {
  type: typeof QUESTIONS_FROM_FILE_ADD;
  questions: Question[];
}

interface QuestionApprovedAction {
  type: typeof QUESTION_APPROVE;
  question: Question;
}

interface QuestionEditedAction {
  type: typeof QUESTION_EDIT;
  question: Question;
}

interface QuestionRemovedAction {
  type: typeof QUESTION_REMOVE;
  question: Question;
}

interface VoteLikeAction {
  type: typeof VOTE_LIKE;
  question: Question;
}

interface VoteDislikeAction {
  type: typeof VOTE_DISLIKE;
  question: Question;
}

export const addQuestions = (questions: Question[], count = 0, pages = 0) => ({
  type: QUESTIONS_ADD,
  questions,
  count,
  pages,
});

export const questionGot = (question: Question) => ({
  type: QUESTION_GET,
  question,
});

export const questionAdded = (question: Question) => ({
  type: QUESTION_ADD,
  question,
});

export const questionsFromFileAdded = (questions: Question[]) => ({
  type: QUESTIONS_FROM_FILE_ADD,
  questions,
});

export const questionApproved = (question: Question) => ({
  type: QUESTION_APPROVE,
  question,
});

export const questionEdited = (question: Question) => ({
  type: QUESTION_EDIT,
  question,
});

export const questionRemoved = (question: Question) => ({
  type: QUESTION_REMOVE,
  question,
});

export const voteLike = (question: Question) => ({
  type: VOTE_LIKE,
  question,
});

export const voteDislike = (question: Question) => ({
  type: VOTE_DISLIKE,
  question,
});

export const getQuestionInterface = () => () => api.questions.getInterface();

export const getQuestions = (
  page = 1,
): ThunkAction<void, AppState, null, Action<string>> => dispatch =>
  api.questions.getQuestions(page).then(({ questions, count, pages }) => {
    dispatch(addQuestions(questions, count, pages));

    return { count, pages };
  });

export const getTopQuestions = (): ThunkAction<
  void,
  AppState,
  null,
  Action<string>
> => dispatch =>
  api.questions.getTop().then(questions => dispatch(addQuestions(questions)));

export const getNotVerifiedQuestions = (): ThunkAction<
  void,
  AppState,
  null,
  Action<string>
> => dispatch =>
  api.questions
    .getNotVerified()
    .then(questions => dispatch(addQuestions(questions)));

export const getQuestionsByFilter = (
  filter: string,
  tag = '',
): ThunkAction<void, AppState, null, Action<string>> => dispatch =>
  api.questions.getByFilter(filter, tag).then(({ tags, questions }) => {
    dispatch(addQuestions(questions));

    return { tags, questions };
  });

export const getQuestionsByQList = (
  qlist: string,
): ThunkAction<void, AppState, null, Action<string>> => dispatch =>
  api.questions
    .getByQList(qlist)
    .then(questions => dispatch(addQuestions(questions)));

export const getQuestionById = (
  id: string,
): ThunkAction<void, AppState, null, Action<string>> => dispatch =>
  api.questions.getById(id).then(question => dispatch(questionGot(question)));

export const getQuestionBySlug = (
  slug: string,
): ThunkAction<void, AppState, null, Action<string>> => dispatch =>
  api.questions
    .getBySlug(slug)
    .then(question => dispatch(questionGot(question)));

export const getQuestionsByAuthor = (
  username: string,
): ThunkAction<void, AppState, null, Action<string>> => dispatch =>
  api.questions
    .getByAuthor(username)
    .then(questions => dispatch(addQuestions(questions)));

export const addQuestion = (
  question: Question,
): ThunkAction<void, AppState, null, Action<string>> => dispatch =>
  api.questions
    .add(question)
    .then(question => dispatch(questionAdded(question)));

export const addQuestionsFromFile = (
  questions: Question[],
): ThunkAction<void, AppState, null, Action<string>> => dispatch =>
  api.questions
    .addFromFile(questions)
    .then(questions => dispatch(questionsFromFileAdded(questions)));

export const approveQuestion = (
  id: string,
): ThunkAction<void, AppState, null, Action<string>> => dispatch =>
  api.questions
    .approve(id)
    .then(question => dispatch(questionApproved(question)));

export const editQuestion = (
  data: any,
): ThunkAction<void, AppState, null, Action<string>> => dispatch =>
  api.questions.edit(data).then(question => dispatch(questionEdited(question)));

export const editQuestionField = (
  id: string,
  field: string,
  value: string,
): ThunkAction<void, AppState, null, Action<string>> => dispatch =>
  api.questions.editField(id, field, value).then(question => {
    dispatch(questionEdited(question));

    return question;
  });

export const removeQuestion = (
  id: string,
): ThunkAction<void, AppState, null, Action<string>> => dispatch =>
  api.questions
    .remove(id)
    .then(question => dispatch(questionRemoved(question)));

export const getSearchedQuestions = (
  query: string,
): ThunkAction<Promise<any>, AppState, null, Action<string>> => dispatch =>
  api.questions.getSearched(query).then(questions => {
    if (questions.length) {
      dispatch(addQuestions(questions));
    }

    return questions;
  });

export const voteQuestion = (
  question: Question,
  action: string,
  userId: string,
): ThunkAction<void, AppState, null, Action<string>> => dispatch => () =>
  api.questions
    .vote(question, action, userId)
    .then(question =>
      action === 'like'
        ? dispatch(voteLike(question))
        : dispatch(voteDislike(question)),
    );

export type QuestionsActionTypes =
  | AddQuestionsAction
  | QuestionGotAction
  | QuestionAddedAction
  | QuestionsFromFileAddedAction
  | QuestionApprovedAction
  | QuestionEditedAction
  | QuestionRemovedAction
  | VoteLikeAction
  | VoteDislikeAction;
