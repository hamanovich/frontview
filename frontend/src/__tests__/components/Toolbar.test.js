import React from 'react';
import { shallow } from 'enzyme';

import { Toolbar } from '../../components/shared/Toolbar';
import { question } from '../../__fixtures__/questions';
import { qlists } from '../../__fixtures__/qlists';

describe('<Toolbar/>', () => {
  const props = {
    user: {},
    question,
    qlists,
    qlistAddQuestion: jest.fn(),
    voteQuestion: jest.fn(),
  };

  const component = shallow(<Toolbar {...props} />);

  it('renders <Toolbar /> component', () => {
    expect(component).toMatchSnapshot();
  });

  it('checks Qlists dropdown existance', () => {
    expect(component.find('DropdownButton')).toHaveLength(1);
  });
});
