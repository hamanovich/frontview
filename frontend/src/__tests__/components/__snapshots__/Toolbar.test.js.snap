// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Toolbar/> renders <Toolbar /> component 1`] = `
<ForwardRef(Bootstrap(ButtonToolbar))>
  <ForwardRef(Bootstrap(ButtonGroup))
    size="sm"
  >
    <ForwardRef(Bootstrap(Button))
      active={false}
      variant="success"
    >
      <FontAwesome
        name="thumbs-up"
      />
       
      1
    </ForwardRef(Bootstrap(Button))>
    <ForwardRef(Bootstrap(Button))
      active={false}
      variant="danger"
    >
      <FontAwesome
        name="thumbs-down"
      />
       
      0
    </ForwardRef(Bootstrap(Button))>
  </ForwardRef(Bootstrap(ButtonGroup))>
  <ForwardRef(Bootstrap(ButtonGroup))
    size="sm"
  >
    <DropdownButton
      id="qlist"
      size="sm"
      title="Add to QList"
      variant="info"
    >
      <ForwardRef(ContextTransform)
        eventKey={0}
        key="5bd6e86025ccd523e4595ad4"
      >
        Js core
      </ForwardRef(ContextTransform)>
      <ForwardRef(Bootstrap(DropdownDivider)) />
      <LinkContainer
        activeClassName="active"
        exact={false}
        replace={false}
        strict={false}
        to="/me/qlist/create"
      >
        <ForwardRef(ContextTransform)>
          Create a new QList
        </ForwardRef(ContextTransform)>
      </LinkContainer>
    </DropdownButton>
  </ForwardRef(Bootstrap(ButtonGroup))>
</ForwardRef(Bootstrap(ButtonToolbar))>
`;
